docs/posts/pydata-performance/index.qmd

```diff
@@ -0,0 +1,696 @@
+import os
+
+smallest_file = min(glob.glob("/data/pypi-parquet/*.parquet"), key=os.path.getsize)
+
+%time df = pd.read_parquet(smallest_file)
+df
+```
+
+::: {.callout-note}
+# I am aware `df` isn't a great variable name.
+:::
+
+Loading the smallest file from the dataset is already pretty close
+to the time it took Ibis and DuckDB to execute the *entire query*.
+
+Let's give Pandas a leg up tell it what columns to use to avoid reading in
+a bunch of data we're not going to use. We can figure these columns out by
+manually inspecting the ibis code above:
+
+```{python}
+columns = ["path", "uploaded_on", "project_name"]
+
+%time df = pd.read_parquet(smallest_file, columns=columns)
+df
+```
+
+Sweet, read times improved!
+
+Let's looks at the memory usage of the `DataFrame`.
+
+```{python}
+df.memory_usage().sum() / (1 << 20)  # MiB
+```
+
+I still have plenty of space to do my analysis, nice!
+
+First, filter the data:
+
+```{python}
+%%time
+df = df[
+    (
+        df.path.str.contains(r"\.(?:asm|c|cc|cpp|cxx|h|hpp|rs|[Ff][0-9]{0,2}(?:or)?|go)$")
+        & ~df.path.str.contains(r"(?:^|/)test(?:|s|ing)|/site-packages/")  # <1>
+    )
+]
+df
+```
+
+1. I altered the original query here to avoid creating an unnecessary
+   intermediate `Series` object.
+
+We've blown **way** past our ~30s Ibis + DuckDB latency budget.
+
+Let's keep going!
+
+Next, group by and aggregate:
+
+```{python}
+#| error: false
+%%time
+df = df.groupby(
+    [
+        df.uploaded_on.dt.floor("M"),
+        df.path.str.extract(r"\.([a-z0-9]+)$", 0),
+    ]
+)
+df
+```
+
+Here we hit the first API issue going back to an [old pandas
+issue](https://github.com/pandas-dev/pandas/issues/15303): we can't truncate
+a timestamp column to month frequency.
+
+Let's try the solution recommended in that issue.
+
+```{python}
+#| error: false
+%%time
+df = df.groupby(
+    [
+        df.uploaded_on.dt.to_period("M").dt.to_timestamp(),
+        df.path.str.extract(r"\.([a-z0-9]+)$", 0)
+    ]
+)
+df
+```
+
+That seems to have gotten past the `floor` issue but now I get an error about
+the shape of the grouping keys. I suspect this is because `extract` returns
+a `DataFrame`. Such is life.
+
+Let's try a different strategy: compute the columns in an `assign` call and
+then do the group by using string column names:
+
+```{python}
+%%time
+df = (
+    df.assign(
+        month=lambda t: t.uploaded_on.dt.to_period("M").dt.to_timestamp(), # <1>
+        ext=lambda t: t.path.str.extract(r"\.([a-z0-9]+)$", 0), # <1>
+    )
+    .groupby(["month", "ext"])
+    .agg({"project_name": lambda s: list(set(s))})
+    .rename(columns={"project_name": "projects"})
+)
+df
+```
+
+1. Later we'll see that Dask doesn't support lambda functions in `assign` and
+   instead we'll create an intermediate `DataFrame` for the `assign` call.
+
+Sort the values, add a new column and do the final aggregation:
+
+```{python}
+%%time
+df = (
+    df.reset_index()  # <1>
+    .sort_values("month", ascending=False)
```

Comment:
Why would you do this before grouping? GroupBy sorts anyway

Reply:
Did you read the rest of the post 😂

Reply:
I mention later that this sorting here is kind of pointless. It's intentionally here though so I plan on leaving it.

Reply:
> I mention later that this sorting here is kind of pointless. It's intentionally here though so I plan on leaving it.

Yeah I found that too, that's why I added 

> If yes, then please ignore my comments to that effect

to the top comment. It just confused me (as a reader) that the query was so inefficient. 


As I said, feel free to ignore if this is intentional